File: codes/backtraking/93.restore-ip-addresses/restore-ip-addresses.cpp
#include <string>
#include <vector>
using namespace std;
class Solution {
    vector<string> result;
    string path;
    using sit = string::const_iterator;

public:
    vector<string> restoreIpAddresses(string s) {
        backtracing(s.cbegin(), s.cend(), 4);
        return result;
    }
    void backtracing(sit beg, sit end, int layer) {
        if (layer == 0) {
            if (beg == end) {
                result.push_back(path);
                // remove the last '.' in result, can't do it in path.
                result.back().pop_back();
            }
            return;

        } else if (end - beg < layer) {
            return;
        }

        for (int i = 1; i <= 3 && beg + i <= end; i++) {
            auto cur = string(beg, beg + i);
            // judge cur is valid
            if (cur.front() == '0' && cur.size() > 1) {
                return;
            }
            int num = 0;
            for (int j = 0; j < cur.size(); ++j) {
                if (isdigit(cur[j])) {
                    num = num * 10 + cur[j] - '0';
                }
            }
            if (num > 255) {
                return;
            }
            path.append(cur);
            path.push_back('.');
            backtracing(beg + i, end, layer - 1);
            path = path.substr(0, path.size() - cur.size() - 1);
        }
    }
};

File: codes/backtraking/93.restore-ip-addresses/restoreIpAddresses.go
package main

import (
	"strconv"
)

func restoreIpAddresses(s string) []string {
	var result []string
	var path []byte

	isValid := func(s string) bool {
		if len(s) < 0 || len(s) > 3 {
			return false
		}
		if len(s) > 1 && s[0] == '0' {
			return false
		}
		if num, err := strconv.Atoi(s); err != nil || num > 255 {
			return false
		}
		return true
	}

	var backtracing func(int, int)

	backtracing = func(beg int, layer int) {
		if layer < 0 {
			return
		}
		if layer == 0 {
			if beg >= len(s) {
				result = append(result, string(path[:len(path)-1]))
				return
			}
		}

		for i := beg; i < beg+3 && i < len(s); i++ {
			if isValid(s[beg : i+1]) {
				path = append(path, s[beg:i+1]...)
				path = append(path, '.')
				backtracing(i+1, layer-1)
				path = path[:len(path)-(i-beg+1)-1]
			} else {
				return
			}
		}

	}
	backtracing(0, 4)
	return result
}
